### PR TITLE
Add `karma-jasmine-html-reporter` 

### PR DESCRIPTION
**Pull request by ebogo1**
This should address some of the concerns in https://github.com/CesiumGS/cesium/issues/10200. The `kjhtml` reporter adds a clickable list of tests after they finish running. Since the Karma Chrome window auto-closes, this is mostly useful when running tests with `npm run test -- --debug`. 

One thing I wasn't sure about was how this interacts with headless Chrome on CI - I ran CI a couple times and the build time seemed reasonable but it's possible it's more correct to disable the html spec reporter on CI runs.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @ebogo1!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/html-reporter/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.

**Comment by ebogo1** [comment]
@ggetz Sounds good - I added a `--html` flag that will push the html reporter to the karma config if provided. Should be good for another look.

File: Specs/customizeJasmine.js [hunk 1]
```diff
@@ -126,6 +126,7 @@ function customizeJasmine(
     window.webglStub = true;
   }
 
+  env.hideDisabled(true);
   //env.catchExceptions(true);
```
Comment (ggetz) [inline Specs/customizeJasmine.js:130]: I know this is unrelated, but's let's go ahead and remove this line.

**Review by ggetz — COMMENTED** [review]
@ebogo1 

* You mentioned there's an option to disable the dots, right? Can we disable them by default? This seem a bit silly since we have so many tests:

image

* Also is it possible to show the entire spec list rather than just the failing specs? I believe that's a big appeal of the html reporter.

* Does it make sense to use `--html` ever _without_ the `--debug` option? It doesn't seem like there is much benefit. If so, we should probably update the documentation to just say `npm run test -- --html --debug` by default.

* We are hiding a lot of these deprecation warnings when running from the command line due to suppressing much of the output. Not for this PR, we'll probably want to clean these up soon, especially if we want to upgrade to the latest jasmine.

image

**Comment by mramato** [comment]
> Since the Karma Chrome window auto-closes, 

This is because we tell it to auto-close, we also tell it to run the tests as soon as it opens.  All of these things are configurable.

**Comment by ebogo1** [comment]
> This is because we tell it to auto-close, we also tell it to run the tests as soon as it opens. All of these things are configurable.

@mramato Currently the `--debug` flag added in https://github.com/CesiumGS/cesium/pull/10170 disables auto-close with `singleRun`. That's why using the HTML reporter without the debug option might not make much sense (per @ggetz's [comment](https://github.com/CesiumGS/cesium/pull/10238#pullrequestreview-923216003)). 

In [db28d2e](https://github.com/CesiumGS/cesium/pull/10238/commits/db28d2e1b8e8f5aa88d0bbc0e0399191b8aaaa12) I added a `"ChromeDebug"` launcher to our Karma config to open the `/debug.html` page on launch (requested by @IanLilleyT [here](https://github.com/CesiumGS/cesium/issues/10200#issuecomment-1081993732)). The `--debug` flag will always enable the HTML reporter and use the new launcher.

Other loose ends - 

> You mentioned there's an option to disable the dots, right? Can we disable them by default? 

I have not found a clean way to do this since the dots come from a CSS file in the node module folder..  

> Also is it possible to show the entire spec list rather than just the failing specs? I believe that's a big appeal of the html reporter.

This took me too long to figure out but the "Spec List" text at the bottom is a button that lets you toggle between the full list and failures:

image

From Ian's comment in the related issue: 

> Have Karma watch files automatically instead of doing npm run build whenever there are code changes. That way I can keep the Karma window open and refresh the page whenever I make changes.

Karma has `autoWatch` and `restartOnFileChange` flags ([docs](http://karma-runner.github.io/6.3/config/configuration-file.html)) but both of these cause the every enabled test suite to re-run. I haven't been able to get code changes to show up in the Karma window by refreshing it yet and am not sure if it's possible.

**Comment by mramato** [comment]
@ebogo1 I don't have all the answers off the top of my head, but I'm confident there is a good solution that works for all of our needs here. What I think is missing is a clear definition of what our goals are for unit testing overall. What do we want the test and debugging experience to be like?

I feel like we are just trying to glob onto the existing jasmine/karma configuration and what we really need to do is take a step back, make sure we understand how all of the tools interact with each other, and then re-implement the karma/jasmine configuration from scratch, deferring to the "old" configuration as a reference.  We should be able to clearly explain why every option is there and what its effect is. 

For example, in the extreme case, `jasmine-browser-runner` did not exist when we added karma. It's how jasmine recommends you run browser-based tests.  Have we evaluated it at all as a replacement for karma?

**Comment by ebogo1** [comment]
@mramato (and as a note to anyone following this PR) - `jasmine-browser-runner` depends on Jasmine 4.0. Upgrading was something we planned on doing anyway so I'm now testing a different branch to resolve the deprecation warnings we see on 3.99 (almost all of them are to do with updating our custom equality matcher syntax). If updating to 4.0 is easy enough, I'll open a PR for the update and try out `jasmine-browser-runner` in earnest. If that works out then I'll close this PR and focus on the other one.

> What I think is missing is a clear definition of what our goals are for unit testing overall. What do we want the test and debugging experience to be like?

There's some urgency to this since the current setup has some extra developer friction compared to the old SpecRunner (https://github.com/CesiumGS/cesium/issues/10200), but our main goals are to have: 
- A quick way to filter specs by name and suite (with similar filtering as the previous web runner had, ideally without having to write/maintain our own comparison code)
- An intuitive process for stepping through unit tests in a debugger (we plan on maintaining a VSCode task to do so, but we’d also like to support a similar workflow via the browser as before as an alternative)
- An easy way to browse all tests visually

**Comment by ggetz** [comment]
Before we removed the old spec runner, [we made sure to ask the community about the implications](https://community.cesium.com/t/changes-to-unit-test-debugging-in-cesiumjs/17537/8) including the change to debug workflow and filtering. After it was removed, it looks like this came down to being too drastic a change in developer workflow preference. If we're not able to use the web runner to address the above points, we should be able to restore the standalone jasmine runner (and we should be able to use a more up-to-date version).

**Pull request closed without merging**